### ISSUE TITLE
neutron port-update updates unnecessary attributes (allowed_address_pairs)

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-openstack#24]
_This issue was originally opened by @michalmedvecky as hashicorp/terraform#15220. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version

v0.9.6

### Affected Resource(s)
- openstack_networking_port_v2

I have a network port in OpenStack which needs a new security group to be attached on it. When I run `terraform plan`, there is a correct output (removing one security group, adding another one):

```
openstack_networking_port_v2.otplb_port_backend.0: Modifying... (ID: f8c309c7-0c81-404d-a2df-41d0e5f5dce7)
  security_group_ids.1691341947: "721c89fb-b8a6-46dd-b031-73038443ded8" => "721c89fb-b8a6-46dd-b031-73038443ded8"
  security_group_ids.1894619809: "ae9af23e-a5d7-45be-a561-21220c74493b" => "ae9af23e-a5d7-45be-a561-21220c74493b"
  security_group_ids.4281792489: "a9f112b3-76f5-4d6b-b573-929df4169a63" => ""
  security_group_ids.691798261:  "" => "2e801289-8912-446a-9bbc-030e498e31fa"
```

but `terraform apply` fails, wanting to unnecessarily update `allowed_address_pairs`, which I have no permissions to update:

```
* openstack_networking_port_v2.otplb_port_backend.0: Error updating OpenStack Neutron Network: Expected HTTP response code [200 201] when accessing [PUT http://1.2.3.4:9696/v2.0/ports/f8c309c7-0c81-404d-a2df-41d0e5f5dce7], but got 403 instead
{"NeutronError": {"message": "(rule:update_port and rule:update_port:allowed_address_pairs) on {'status': u'ACTIVE', u'allowed_address_pairs': [], 'attributes_to_update': [u'allowed_address_pairs', u'security_groups'], 'network_id': u'25554292-3fcb-4063-900c-6aacee863f3c', 'tenant_id': u'3e8f42a379284709b3abd556a6885102', 'id': u'f8c309c7-0c81-404d-a2df-41d0e5f5dce7', u'security_groups': [u'721c89fb-b8a6-46dd-b031-73038443ded8', u'ae9af23e-a5d7-45be-a561-21220c74493b', u'2e801289-8912-446a-9bbc-030e498e31fa'], u'network:tenant_id': u'97ca9f15cdcb4e65b1a52f8fd90a1f04'} by {'domain': None, 'project_name': u'sec_iam_dev', 'tenant_name': u'sec_iam_dev', 'project_domain': None, 'timestamp': '2017-06-09 14:35:21.544773', 'auth_token': 'c9755c1833464e78af3145bbaed45705', 'resource_uuid': None, 'is_admin': False, 'user': u'11c0fd1d2a3a475aaeb31f1f637c2f91', 'tenant': u'3e8f42a379284709b3abd556a6885102', 'read_only': False, 'project_id': u'3e8f42a379284709b3abd556a6885102', 'user_id': u'11c0fd1d2a3a475aaeb31f1f637c2f91', 'show_deleted': False, 'roles': [u'Member'], 'user_identity': '11c0fd1d2a3a475aaeb31f1f637c2f91 3e8f42a379284709b3abd556a6885102 - - -', 'tenant_id': u'3e8f42a379284709b3abd556a6885102', 'request_id': 'req-6c9da58b-8d9d-46f6-abac-fc32a9026076', 'user_domain': None, 'user_name': u'mmedvecky'} disallowed by policy", "type": "PolicyNotAuthorized", "detail": ""}}
```

I would expect terraform to only do update on attributes that are changed. The behavior is anyway wrong; as you can see in the output, it wants to set empty `allowed_address_pairs`, what is incorrect (there are some existing values in the state file).

**Comment by fatmcgav** [comment]
@michalmedvecky Are you able to provide an example resource from the state file that shows the `allowed_address_pairs` value set?

**Comment by michalmedvecky** [comment]
Here you go:
```
                "openstack_networking_port_v2.otplb_port_backend.0": {
                    "type": "openstack_networking_port_v2",
                    "depends_on": [
                        "openstack_networking_network_v2.net-backend",
                        "openstack_networking_secgroup_v2.common",
                        "openstack_networking_secgroup_v2.deployer_access",
                        "openstack_networking_secgroup_v2.otp_lb_backends"
                    ],
                    "primary": {
                        "id": "f8c309c7-0c81-404d-a2df-41d0e5f5dce7",
                        "attributes": {
                            "admin_state_up": "true",
                            "all_fixed_ips.#": "1",
                            "all_fixed_ips.0": "100.127.177.121",
                            "allowed_address_pairs.#": "0",
                            "device_id": "504e91f5-eae4-44a8-abb2-2f30ba7991c5",
                            "device_owner": "compute:sec",
                            "id": "f8c309c7-0c81-404d-a2df-41d0e5f5dce7",
                            "mac_address": "fa:16:3e:9f:8f:dd",
                            "name": "otplb01.domain.redacted",
                            "network_id": "25554292-3fcb-4063-900c-6aacee863f3c",
                            "region": "RegionOne",
                            "security_group_ids.#": "3",
                            "security_group_ids.1691341947": "721c89fb-b8a6-46dd-b031-73038443ded8",
                            "security_group_ids.1894619809": "ae9af23e-a5d7-45be-a561-21220c74493b",
                            "security_group_ids.3214647677": "eb20572f-ea7f-4ef7-9dbb-ee62f4ee5d47",
                            "tenant_id": "3e8f42a379284709b3abd556a6885102"
                        },
                        "meta": {
                            "e2bfb730-ecaa-11e6-8f88-34363bc7c4c0": {
                                "create": 600000000000,
                                "delete": 600000000000
                            }
                        },
                        "tainted": false
                    },
                    "deposed": [],
                    "provider": ""
                },
```

And the actual interface from neutron:

```
+-----------------------+----------------------------------------------------------------------------------------+
| Field                 | Value                                                                                  |
+-----------------------+----------------------------------------------------------------------------------------+
| admin_state_up        | True                                                                                   |
| allowed_address_pairs |                                                                                        |
| binding:vnic_type     | normal                                                                                 |
| created_at            | 2017-03-26T21:48:05                                                                    |
| description           |                                                                                        |
| device_id             | 504e91f5-eae4-44a8-abb2-2f30ba7991c5                                                   |
| device_owner          | compute:sec                                                                            |
| dns_name              | otplb01.domain.redacted                                                  |
| extra_dhcp_opts       |                                                                                        |
| fixed_ips             | {"subnet_id": "30cdd2c8-a696-49de-b45c-7fb162c732a7", "ip_address": "100.127.177.121"} |
| id                    | f8c309c7-0c81-404d-a2df-41d0e5f5dce7                                                   |
| mac_address           | fa:16:3e:9f:8f:dd                                                                      |
| name                  | otplb01.domain.redacted                                                  |
| network_id            | 25554292-3fcb-4063-900c-6aacee863f3c                                                   |
| security_groups       | 721c89fb-b8a6-46dd-b031-73038443ded8                                                   |
|                       | a9f112b3-76f5-4d6b-b573-929df4169a63                                                   |
|                       | ae9af23e-a5d7-45be-a561-21220c74493b                                                   |
| status                | ACTIVE                                                                                 |
| tenant_id             | 3e8f42a379284709b3abd556a6885102                                                       |
| updated_at            | 2017-05-30T12:11:16                                                                    |
+-----------------------+----------------------------------------------------------------------------------------+
```

**Comment by fatmcgav** [comment]
@michalmedvecky So looking at the state and port output above, the `allowed_address_pairs` value is actually empty... 

Looking at the code, it looks like `allowed_address_pairs` is always being provided on an update - see [openstack_networking_port_v2#239](https://github.com/terraform-providers/terraform-provider-openstack/blob/f9128ce6f765bfd17d3214ce758fd0a72f1c5a6f/openstack/resource_openstack_networking_port_v2.go#L239-L246).
It looks like this behaviour was changed in https://github.com/hashicorp/terraform/pull/13604, which was related to https://github.com/hashicorp/terraform/issues/13531 that I raised, due to failures with passing a `null` value...  

So it looks like we've got conflicting behaviour on different OS cloud providers :( 

@jtopjian Thoughts?

**Comment by jtopjian** [comment]
Ah yes, I recall this one.

What if we reverted the behavior to not pass `allowed_address_pairs`, which seems to be the standard behavior(?) and try catching an error (whichever error you're seeing). If an error is caught, try passing both?

**Comment by fatmcgav** [comment]
Yeh, I guess that's one option... 

The other one I thought of is doing similar to the changes made for FWaaS router insertion, whereby a `no_routers` option was added, to explicitly signify that `[]` should be passed... Maybe something similar here might be a better solution?

**Comment by jtopjian** [comment]
I think the difference is that `no_routers` is applicable to everyone and it has to do with the core Neutron API (unless I'm mistaken?). I think the fix implemented in https://github.com/hashicorp/terraform/pull/13604 was to handle a specific cloud, but since my testing showed it was OK to pass both `allowed_address_pairs` and `security_groups`, we thought it'd be safe to do globally. But, I have a feeling that @michalmedvecky's cloud has a `policy.json` setting to prohibit him from setting `allowed_address_pairs`. Since https://github.com/hashicorp/terraform/pull/13604 was a deviation from the norm, it's now affecting a user it otherwise wouldn't have if the patch wasn't applied. whew :)

This sounds like it goes back to one of our discussions on how to handle oddities found in specific clouds. I don't like the idea of having first-class arguments to flag a unique action. However, I think it's acceptable to re-use `value_specs` or introduce either a provider-level or resource-level `hacks` map where we can look for specific flags. In this case, something like `pass_allowed_address_pairs_on_update = true`.

Thoughts?

**Comment by fatmcgav** [comment]
I think that's an accurate description of the issue... ;) 

It is feeling more and more like we need some way to handle _odd_ behaviour for specific vendors... 
Having a `vendor_modifications` map at the provider may well be the best approach... At-least if it's at that level it can be easily documented, and also less sprawl across the various resources... 

I'll see if I can find some time to flesh the idea out a bit...

**Comment by michalmedvecky** [comment]
Guys ... let us breathe! 😕

**Comment by fatmcgav** [comment]
@michalmedvecky Essentially, the intention is to revert the changes made in https://github.com/hashicorp/terraform/pull/13604, and then implement a _better_ solution to the problem raised in https://github.com/hashicorp/terraform/issues/13531...

**Comment by jtopjian** [comment]
Fixed by #114. A release with this fix will be out soon.
